torch/_functorch/aot_autograd.py

```diff
@@ -1803,7 +1803,8 @@ def compiled_function(*args):
                             updated_inpt.stride(),
                             updated_inpt.storage_offset(),
                         )
-                    original_inpt.copy_(updated_inpt)
+                    with torch.no_grad():
+                        original_inpt.copy_(updated_inpt)
```

Comment:
It's not obvious to me why disabling autograd here is correct. The old code you could backprop through subsequent uses of input; with the new code you can't, right?

Reply:
oh you're totally right. I think I should do something like:
```
if original_inpt.is_leaf:
    ctx = torch.no_grad()
else:
    ctx = nullcontext()

with ctx:
    original_inpt.copy_(updated_inpt)
```

Reply:
and CI right now is correctly failing, for the reason you described

Reply:
(updated, I see tests passing locally)

Reply:
Mmmf, I am willing to accept this to unblock but it still seems fishy. If no grad was still set on a subsequent call, then this shouldn't have been necessary. But if the no grad context changed, we in fact need to guard fail and recompile, no?

Reply:
hmmm.... I tried to repro the example that you gave, and it looks to me like dynamo is guarding properly. Unless I'm misunderstanding?
```

# I confirmed that this repro compiles twice - indicating the dynamo recompiles on the `torch.no_grad()` context change?
@torch.compile(backend="aot_eager")
def f(a):
    a.clone().add_(1)
    return a

inp = [torch.ones(3, requires_grad=True)]
with torch.no_grad():
    out = f(*inp)
out = f(*inp)
```

```
# I confirmed that this compiles once, and then erorrs out on the second call of f(), where we are no longer in no_grad mode.
@torch.compile(backend="aot_eager")
def f(a):
    a.add_(1)
    return a

inp = [torch.ones(3, requires_grad=True)]
with torch.no_grad():
    out = f(*inp)
out = f(*inp)
```

Reply:
So then why do we need to patch it here 🤔

Reply:
the case that this PR fixes is when the no_grad() is **inside** the function we're compiling:

```
@torch.compile(backend='aot_eager')
def f(x):
    with torch.no_grad():
        x.add_(1)

x = torch.zeros(2, requires_grad=True)
# Fails with "RuntimeError: a leaf Variable that requires grad is being used in an in-place operation"
out = f(x)
```

Reply:
Offline discussion: we need to no grad even if we mutate a non-leaf tensor, so Brian is going to try to track this directyl.

Reply:
I think the other edge case I'm worried about is something like this:

```
x.add_(y)
with no_grad():
    x.add_(y2)
x.add_(y3)
```

It's not clear to me that we can actually express this as a single `copy_` outside of the custom autograd function anymore...

Reply:
@albanD convinced me this is fine, you just have to make sure you have a funky grad for x which only has the backwards contribution from y and y3 but not y2.

Reply:
So for this particular example, you want the backward to be

```
grad_x = grad_y + grad_y3
```

as the y2 addition being untracked due to the no grad. So all you need to do is make sure *this* is what goes in your backward graph, and then it's fine to set x's grad_fn for this.